### ISSUE TITLE
[System.Net.NetworkInformation] Follow up on miscellaneous perf / style issues

### ISSUE THREAD

**Issue by mellinoe** [dotnet/corefx#4060]
- [x] There are a number of // PERF markers in the code
- [x] @pgavlin suggested that, when collecting data using sysctl on OSX, we could try to use a stack-allocated buffer of some fixed size before falling back to heap allocation if the call fails. There are a number of instances of this in pal_networkstatistics.cpp.
- [x] A number of properties on various types in the library can be lazily instantiated, assuming that doing so will not alter the resulting object that is created. I don't have an exhaustive list of all the properties that can be optimized this way. We should match the behavior of Windows here with regards to when the data is finalized for those properties.
- [x] When parsing the Linux network files, we should probably use invariant string comparisons, rather than the default culture-aware.
- [x] Address miscellaneous comments from the previous PR (https://github.com/dotnet/corefx/pull/4039)
- [x] Move TcpState mapping from C# to native shim
- [x] Consider bailing on sysctl calls if estimated buffer size isn't big enough. (Most code examples do this and I've never seen the estimated size be incorrect).

**Comment by mellinoe** [comment]
Regarding bullet 3: I have been looking for additional places to lazily instantiate and haven't seen anything jumping out at me. I'll mark it as complete and can address things individually if they become a perf issue.

**Comment by mellinoe** [comment]
I also investigated the second bullet point, but came across some interesting / confusing behavior with regards to how sysctl is functioning on OSX. The function doesn't seem to be returning the values it is documented to return (0 or -1 with errno), so it is difficult to dynamically expand a buffer (or go from stack allocation to heap) given how the function is performing. Additionally, all of the usage of sysctl that I have found on the web has been analogous to our usage, which is to query the buffer size first, and then call the function again with a buffer of that size. Most implementations also just return an error code if the second call fails, so we might even consider just doing that.
